### ISSUE TITLE
OS X Different Notification Icon Per Service 

### ISSUE THREAD

**Issue by billyfordon** [getferdi/ferdi#793]
Is it possible to use the actual service icon in the OS X notification rather than the Ferdi icon? It is not a must-have, but definitely a very nice to have as it would make it very clear what service is sending the notification.

**Comment by billyfordon** [comment]
Browsing through the code, it seems like it would work simply by removing this check: https://github.com/getferdi/ferdi/blob/6999476b9eb71880e16b413f27a7df5bf458cd47/src/stores/AppStore.js#L282

I am far from being an expert on this, but is this even necessary? What was the reason for removing blobs? Do we really care if the blob gets stored on disk?

**Comment by vantezzen** [comment]
As far as I can tell, blobs cannot be used for notification icons.

The notifications constructor (https://developer.mozilla.org/en-US/docs/Web/API/Notification/icon) requires `the URL of an icon` but blobs URLs are probably too scoped to the current WebView and can't be used system-wide. That's why we need to have the icon stored on disk. We may be able to use the icons from the uncompressed recipe directory (`~/Library/Application Support/Ferdi/recipes`) though.

**Comment by billyfordon** [comment]
Yes, I was doing some reading and it looks like that is a constraint. However, even going one step further by allowing the notification to use the contact icon (e.g., the picture of the person or group sending the message) should be possible, no? Since the web application must give you an URL or a blob for each contact to display it in the browser, we could just save it to disk and display it, no?

**Comment by vantezzen** [comment]
This requires Ferdi to know where to find the picture for each notification, which is different for each service and can be quite hard to decide depending on the service. This would also have the same problem as the current solution in that blobs would need to be transferred to disk first.

**Comment by billyfordon** [comment]
True. I know of a couple OS X apps (Rambox, One Chat) that do this, so it should not be too hard... What about an 80-20 approach? Do this for the main services (WhatsApp, Telegram, Skype, etc.) and leave the others be. Happy to contribute some code if that would be acceptable...

**Comment by iliyang** [comment]
Agreed with @billyfordon. I'm coming from Rambox and so far the only feature I miss from it is the per-service notification icon – makes it very easy to see if a notification is important (e.g. coming from the work Slack) or not (e.g. from the wife via Messenger 😇 ).

**Comment by martinbernat** [comment]
I'm playing with this right now and I've managed this result under macOS:
_(the electron icon on the left is because of dev build, imagine Ferdi icon instead)_
![image](https://user-images.githubusercontent.com/6809891/142486685-ae02cdb7-e3d8-4944-ab00-39de942adcf8.png)

The code change is pretty simple, just 3 lines of code. The only problem is that Electron does not accept SVG icons, only PNG. I'm getting the PNG icon dynamically from https://cdn.franzinfra.com/recipes/dist/{RECIPE}/src/icon.png - found that URL in the Ferdi code. Are we allowed to use that (Franzinfra)? Can we have copies somewhere under getferdi domain, or https://cdn.jsdelivr.net/gh/getferdi...? We could  maybe also cache the icons locally after downloading.

Another way could be convert our SVGs to PNG dynamically, there are some JS libraries to do that I guess.

3rd and worst way is to have both SVG and PNG versions in the recipe package which will inflate the app footprint.

Maybe there is better way please advise @vantezzen , @vraravam - I'm a Ferdi newbie :)

**Comment by vraravam** [comment]
> 3rd and worst way is to have both SVG and PNG versions in the recipe package which will inflate the app footprint.

This actually used to be the case about 6+ months ago. I had removed the png's for the same reason - that the app footprint could be reduced.

Instead of shipping the png icon bundled inside the app, can this use a http url instead? Would that solve both problems? (just that it would require a network connection at least for the first time an icon is requested, but this is a messaging app after-all lol)

This can also tie into this feature request/idea: https://github.com/getferdi/ferdi/discussions/1601

**Comment by vraravam** [comment]
> I'm getting the PNG icon dynamically from

Rereading your comment above: This is great @martinbernat !!!! If I understand correctly, you are able to render with the http url, which then means it doesn't increase the app size with binary png files, is that correct?

**Comment by vraravam** [comment]
> Can we have copies somewhere under getferdi domain, or https://cdn.jsdelivr.net/gh/getferdi...?

This would be even better - since we have some recipes that are not present in franz, and also some with changed names. Tagging @kytwb and @vantezzen since I don't have login into the cdn portion of getferdi

**Comment by martinbernat** [comment]
Yes correct, I can render an URL of service's PNG icon _(now using https://cdn.franzinfra.com/recipes/dist/telegram/src/icon.png)_, and put it into the notification settings directly. No app size increasing. 

If we can have all SVG and PNG icons on the https://cdn.jsdelivr.net/gh/getferdi CDN that would be great! We could then probably get rid of the  https://cdn.franzinfra.com/ dependency.

Now trying to test it on Win/Linux _(+ Gnome versions)_ how  it works there, but I do not expect huge problems since it's a feature of Electron. 
I'm also going to study how caching works in Electron. It would be more sensible to cache icons instead of permanently downloading them from CDN.

**Comment by vraravam** [comment]
That's great @martinbernat ! If you can raise a draft PR, I can then use your fork/branch and test the same on macos.
